velox/exec/HashProbe.cpp

```diff
@@ -1436,18 +1500,342 @@ void HashProbe::setRunning() {
+  VELOX_CHECK(task->pauseRequested());
+  const std::vector<Operator*> operators =
+      task->findPeerOperators(operatorCtx_->driverCtx()->pipelineId, this);
+  bool hasMoreProbeInput{false};
+  for (auto* op : operators) {
+    HashProbe* probeOp = dynamic_cast<HashProbe*>(op);
+    VELOX_CHECK_NOT_NULL(probeOp);
+    VELOX_CHECK(probeOp->canReclaim());
+    if (probeOp->nonReclaimableState()) {
+      RECORD_METRIC_VALUE(kMetricMemoryNonReclaimableCount);
+      ++stats.numNonReclaimableAttempts;
+      LOG(WARNING) << "Can't reclaim from hash probe operators from node pool["
+                   << pool()->parent()->name() << "], usage: "
+                   << succinctBytes(pool()->parent()->currentBytes());
+      return;
+    }
+    hasMoreProbeInput |= !probeOp->noMoreSpillInput_;
+  }
+
+  spillOutput(operators);
+
+  SpillPartitionSet spillPartitionSet;
+  if (hasMoreProbeInput) {
+    VELOX_CHECK_GT(table_->numDistinct(), 0);
+    // Only spill hash table if any hash probe operators still has input probe
+    // data, otherwise we skip this step.
+    spillPartitionSet = spillTable();
+    VELOX_CHECK(!spillPartitionSet.empty());
+  }
+  const auto spillPartitionIdSet = toSpillPartitionIdSet(spillPartitionSet);
+
+  for (auto* op : operators) {
+    HashProbe* probeOp = dynamic_cast<HashProbe*>(op);
+    VELOX_CHECK_NOT_NULL(probeOp);
+    // Setup all the probe operators to spill the rest of probe inputs if the
+    // table has been spilled.
+    if (!spillPartitionSet.empty()) {
+      VELOX_CHECK(hasMoreProbeInput);
+      probeOp->maybeSetupInputSpiller(spillPartitionIdSet);
+    }
+    probeOp->pool()->release();
+  }
+
+  // Clears memory resources held by the built hash table.
+  table_->clear(true);
+  // Sets the spilled hash table in the join bridge.
+  if (!spillPartitionIdSet.empty()) {
+    joinBridge_->setSpilledHashTable(std::move(spillPartitionSet));
+  }
+}
+
+void HashProbe::spillOutput(const std::vector<Operator*>& operators) {
+  struct SpillResult {
+    const std::exception_ptr error{nullptr};
+
+    explicit SpillResult(std::exception_ptr _error) : error(_error) {}
+  };
+
+  std::vector<std::shared_ptr<AsyncSource<SpillResult>>> spillTasks;
+  auto* spillExecutor = spillConfig()->executor;
+  for (auto* op : operators) {
+    HashProbe* probeOp = static_cast<HashProbe*>(op);
+    spillTasks.push_back(
+        std::make_shared<AsyncSource<SpillResult>>([probeOp]() {
+          try {
+            probeOp->spillOutput();
+            return std::make_unique<SpillResult>(nullptr);
+          } catch (const std::exception& e) {
+            LOG(ERROR) << "Spill output from hash probe pool "
+                       << probeOp->pool()->name() << " failed: " << e.what();
+            // The exception is captured and thrown by the caller.
+            return std::make_unique<SpillResult>(std::current_exception());
+          }
+        }));
+    if ((spillTasks.size() > 1) && (spillExecutor != nullptr)) {
```

Comment:
is there any case where spillTasks.size() == 0 ? Same for spillExecutor, can it ever be a nullptr? if not, should we instead replace the if-condition with VELOX_CHECK